docs/ha-considerations.md

```diff
@@ -0,0 +1,337 @@
+
+<!-- TOC -->
+
+- [High Availability Considerations](#high-availability-considerations)
+    - [Overview](#overview)
+    - [Options for Software Load Balancing](#options-for-software-load-balancing)
+        - [keepalived and haproxy example](#keepalived-and-haproxy-example)
+        - [kube-vip example](#kube-vip-example)
+        - [Bootstrap the cluster](#bootstrap-the-cluster)
+
+## Options for Software Load Balancing
+
+When setting up a cluster with more than one control plane, one will typically put the API Server instances behind a load balancer and use the `--control-plane-endpoint` option when running `kubeadm init` for the new cluster to use it.
+
+Of course, the load balancer itself should be highly available, too. This is usually achieved by setting up more load balancers in a cluster of hosts managing a virtual IP, so that always the load balancer on the host currently holding the vIP will be used.
+
+In some environments, like in data centers with dedicated load balancing components, this functionality may already be available. If it is not, software load balancing can be used. In that case some preparation is necessary before bootstrapping a cluster.
+
+Since this is infrastructure Kubernetes runs on and hence is not part of Kubernetes or `kubeadm`, this must be taken care of separately. In the following sections we give examples that have been working for some people while of course there are potentially dozens of other possible configurations.
+
+## keepalived and haproxy example
+
+For providing load balancing from a virtual IP the combination [keepalived](https://www.keepalived.org) and [haproxy](https://www.haproxy.com) has been around for a long time and can be considered well-known and well-tested:
+- The `keepalived` service provides a virtual IP managed by a configurable health check. Due to the way the virtual IP is implemented, all the hosts between which the virtual IP is negotiated need to be in the same IP subnet.
+- The `haproxy` service can be configured for simple stream-based load balancing thus allowing TLS termination to be handled by the API Server instances behind it.
+
+This combination can be run either as services on the operating system or as static pods on the control plane hosts. The service configuration is identical for both cases.
+
+### keepalived configuration
+
+The `keepalived` configuration consists of two files: the service configuration file and a health check script which will be called periodically to verify that the node holding the virtual IP is still operational.
+
+The files are assumed to reside in a _/etc/keepalived_ directory. Note that however some Linux distributions may keep them elsewhere. The following configuration has been successfully used with `keepalived` version 1.3.5:
```

Comment:
_/etc/keepalived_ ->  `/etc/keepalived` (with code back quotes)

Note that however -> Note that